c/cert/test/rules/EXP39-C/test.c

```diff
@@ -0,0 +1,103 @@
+#include <stdlib.h>
+#include <string.h>
+
+void test_incompatible_arithmetic() {
+  float f = 0.0f;
+  int *p = (int *)&f; // NON_COMPLIANT - arithmetic types are not compatible
+                      // with each other
+  (*p)++;
+
+  short s[2];
+  (int *)&s; // NON_COMPLIANT
+
+  (short(*)[4]) & s; // NON_COMPLIANT - array of size 2 is not compatible with
+                     // array of size 4 (n1570 6.7.6.2 paragraph 7)
+  (short(*)[2]) & s; // COMPLIANT
+
+  // char may be signed or unsigned, and so is not compatible with either
+  char c1;
+  (signed char *)&c1;   // NON_COMPLIANT
+  (unsigned char *)&c1; // COMPLIANT - the underlying byte representation is
+                        // always compatible
+  (char *)&c1;          // COMPLIANT - same type
+
+  // int is defined as signed, so is compatible with all the signed versions
+  // (long, short etc. are similar)
+  int i1;
+  (signed int *)&i1;   // COMPLIANT
+  (int *)&i1;          // COMPLIANT
+  (signed *)&i1;       // COMPLIANT
+  (unsigned int *)&i1; // NON_COMPLIANT
+  (const int *)&i1;    // COMPLIANT - adding a const specifier is permitted
+}
+
+struct {
+  int a;
+} * s1;
+struct {
+  int a;
+} * s2;
+struct S1 {
+  int a;
+} * s3;
+struct S1 *s4;
+
+// TODO test across files
```

Comment:
TODO? Do you need to test across files? I think this test should be sufficient.

Reply:
Agreed. This comment is probably left over from earlier development, but test cases across files shouldn't be necessary.